Feature/implement hooks

petervanderdoes opened the pull request:
Pull request for issue #171

petervanderdoes committed:
Add directory setup to the git flow init process

git flow will fail when the directories are not setup in the git config.

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes committed:
Add directory setup to the git flow init process

git flow will fail when the directories are not setup in the git config.

Signed-off-by: Peter van der Does <peter@avirtualhome.com>
Signed-off-by: Vincent Driessen <vincent@3rdcloud.com>

petervanderdoes committed:
Rename the Hooks directory variable.

Renamed so it won't be confused with the already present GITFLOW_DIR variable, which holds the dorectory info of where the gitflow scripts are located.

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes committed:
Rename the Hooks directory variable.

Renamed so it won't be confused with the already present GITFLOW_DIR variable, which holds the dorectory info of where the gitflow scripts are located.

Signed-off-by: Peter van der Does <peter@avirtualhome.com>
Signed-off-by: Vincent Driessen <vincent@3rdcloud.com>

petervanderdoes committed:
Implements do_hook function

Signed-off-by: Peter van der Does <peter@avirtualhome.com>
Signed-off-by: Vincent Driessen <vincent@3rdcloud.com>

petervanderdoes committed:
Implements do_hook function

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes committed:
Add the pre and post hook calls.

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes committed:
Add the pre and post hook calls.

Signed-off-by: Peter van der Does <peter@avirtualhome.com>
Signed-off-by: Vincent Driessen <vincent@3rdcloud.com>

petervanderdoes committed:
Add some documentation to the do_hook function

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes committed:
Add some documentation to the do_hook function

Signed-off-by: Peter van der Does <peter@avirtualhome.com>
Signed-off-by: Vincent Driessen <vincent@3rdcloud.com>

petervanderdoes committed:
Remove trailing whitespace

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes committed:
Remove trailing whitespace

Signed-off-by: Peter van der Does <peter@avirtualhome.com>
Signed-off-by: Vincent Driessen <vincent@3rdcloud.com>

petervanderdoes committed:
Makes the feature more compliant with nvie comments in #171

Renames the called hooks to new proposed naming convention.
Removes the configurable configuration/hooks directory. All hooks are to be stored in .git/hooks

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes committed:
Makes the feature more compliant with nvie comments in #171

Renames the called hooks to new proposed naming convention.
Removes the configurable configuration/hooks directory. All hooks are to be stored in .git/hooks

Signed-off-by: Peter van der Does <peter@avirtualhome.com>
Signed-off-by: Vincent Driessen <vincent@3rdcloud.com>

bobthecow commented:
Lots of whitespace drive-bys here, but other than that, :+1:

c00kiemon5ter commented on gitflow-common:
```diff
@@ -316,3 +317,42 @@ require_branches_equal() {
+
+		return_code=$?
+
+		if [ $return_code -gt 0 ]; then
```
I think `-ne` would be better in place of `-gt`

c00kiemon5ter commented:
use `?w=1` at the end of the url to see a cleaner diff

bobthecow commented:
I know. I just fundamentally disagree with committing whitespace-only
changes to files in someone else's project ;)

nvie committed:
Revert all whitespace changes.

nvie committed:
Refactor the code a bit.

It now has explicit pre and post functions to communicate the difference
in exit code handling.  Also, allow for passing in positional arguments
that make sense on a per-action basis.

nvie committed:
Call the hooks explicitly on a per-action basis, with positional
arguments that are relevant for the action at hand.

nvie committed:
Add some example hook scripts.

nvie commented:
Take a look at this: https://github.com/nvie/gitflow/compare/develop...feature%2Fimplement-hooks

I've currently only implemented it for the `feature` actions, not yet for `hotfix` and `release`, but they will be trivial. What do you guys think? (The files inside the `hooks/` directory serve as example scripts that you can use as a template when writing your own hook scripts.)

petervanderdoes committed:
Merge remote-tracking branch 'nvie/feature/implement-hooks' into feature/implement-hooks

Conflicts:
	gitflow-common

petervanderdoes committed:
Merge remote-tracking branch 'nvie/feature/implement-hooks' into feature/implement-hooks

Conflicts:
	gitflow-common

petervanderdoes committed:
Merge branch 'feature/implement-hooks' of github.com:petervanderdoes/gitflow into feature/implement-hooks

petervanderdoes commented:
Why is there a different hook function for pre and post?

Shouldn't there be a message if the post script hook fails as well?

petervanderdoes commented:
BTW apologies about the white space. My editor was setup to automatically remove eol white space.

nvie commented:
I think it's useful to display a message when the post-hook fails, but I also think there are some cases where errors should be tolerated (e.g. a best-effort "notification" message or something). This puts the post-hook script itself in control as to whether or not errors are acceptable.

petervanderdoes commented:
I agree that the scripts, post and pre, are in control whether or not errors are acceptable. I think that notification of success or  more descriptive failure message is also upon the scripts. In a pre script that means, if the error is acceptable, you exit with exit code 0 and git flow continues happily. In a post script there is no such option, git flow will always continue happily.

Looking at where the post_hooks are called from right now, an example where a die from the git flow script could be useful in a post hook would be in the git flow feature finish.
The hook is called before the finish cleanup is done. In the post script something might be determined where the merged branches shouldn't be deleted.

petervanderdoes committed:
Adds pre/post hooks

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes committed:
Add bare-bone hook scripts for feature and hotfix.

Bare bone pre and post cripts for the hooks of "git flow feature" and "git flow hotfix"

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes committed:
Implements hooks for git flow release

Hooks added
Bare bone scripts for git flow release added
Fixes wrong post hook call for git flow hotfix publish

Signed-off-by: Peter van der Does <peter@avirtualhome.com>

petervanderdoes commented:
I believe I implemented all hooks as described in #171.
If somebody can take a look, I appreciate that.

booch commented:
:+1:

What else needs to happen to get this merged?

petervanderdoes commented:
Not sure,

I decided to implement the feature on my own fork, including other features and bug fixes.
I recently released 1.0-avh, see the [changelog](http://goo.gl/DRNTG) for more details

baby-gnu commented:
Hello,

I tried to use this feature and wonder if it's not possible to run hooks from other places:
- `HOOKS_DIR`: is `.git/hooks`
- `GITFLOW_USER_HOOKS_DIR`: configurable with `gitflow.hooks-dir.user`
- `GITFLOW_SYSTEM_HOOKS_DIR`: configurable with `gitflow.hooks-dir.system`

This will ease the share and updates of pre/post hooks and filters in a team, for example by distributing a new distribution package or managing them in a git repository itself.

Regards.

Update: Implemented in petervanderdoes/gitflow#1

wenzowski commented:
Would love to see this merged. Anything else needed for that to happen?

nvie commented:
Closing all PRs as I'm archiving this repo. git-flow lives on at [git-flow-next](https://github.com/gittower/git-flow-next) if you're still using it. Thanks for contributing!
